orchestrator/careplancontributor/ehr/notifier.go

```diff
@@ -0,0 +1,373 @@
+		if err != nil {
+			return nil, err
+		}
+		bundels = append(bundels, bundle)
+	}
+
+	return &bundels, nil
+}
+
+// putMapListValue adds a value to a map of lists.
+//
+// Parameters:
+//   - refTypeMap: The map to update.
+//   - refType: The key for the map.
+//   - refId: The value to add to the list.
+func putMapListValue(refTypeMap map[string][]string, refType string, refId string) {
+	values := refTypeMap[refType]
+	if values == nil {
+		values = []string{refId}
+	} else if !slices.Contains(values, refId) {
+		values = append(values, refId)
+	}
+	refTypeMap[refType] = values
+}
+
+// findQuestionnaireInputs extracts questionnaire input references from a list of tasks.
+//
+// Parameters:
+//   - tasks: A list of FHIR tasks.
+//
+// Returns:
+//   - []string: A list of questionnaire input references.
+func findQuestionnaireInputs(tasks []fhir.Task) []string {
+	var questionnaireRefs []string
+	for _, task := range tasks {
+		questionnaireRefs = append(questionnaireRefs, fetchTaskInputs(task)...)
+	}
+	return questionnaireRefs
+}
+
+// findQuestionnaireOutputs extracts questionnaire output references from a list of tasks.
+//
+// Parameters:
+//   - tasks: A list of FHIR tasks.
+//
+// Returns:
+//   - []string: A list of questionnaire output references.
+func findQuestionnaireOutputs(tasks []fhir.Task) []string {
+	var questionnaireResponseRefs []string
+	for _, task := range tasks {
+		questionnaireResponseRefs = append(questionnaireResponseRefs, fetchTaskOutputs(task)...)
+	}
+	return questionnaireResponseRefs
+}
+
+// fetchTaskOutputs extracts questionnaire response references from a task's outputs.
+//
+// Parameters:
+//   - task: A FHIR task.
+//
+// Returns:
+//   - []string: A list of questionnaire response references.
+func fetchTaskOutputs(task fhir.Task) []string {
+	var questionnaireResponseRefs []string
+	if task.Output != nil {
+		for _, output := range task.Output {
+			if output.ValueReference != nil &&
+				output.ValueReference.Reference != nil {
+				matchesType := isOfOutputType(output, "QuestionnaireResponse")
+				if matchesType {
+					reference := *output.ValueReference.Reference
+					if !slices.Contains(questionnaireResponseRefs, reference) {
+						questionnaireResponseRefs = append(questionnaireResponseRefs, reference)
+					}
+				}
+			}
+		}
+	}
+	return questionnaireResponseRefs
+}
+
+// isOfOutputType checks if a task output is of a specific type.
+//
+// Parameters:
+//   - output: The task output to check.
+//   - typeName: The type name to check against.
+//
+// Returns:
+//   - bool: True if the output is of the specified type, false otherwise.
+func isOfOutputType(output fhir.TaskOutput, typeName string) bool {
+	matchesType := false
+	if output.ValueReference.Type != nil {
+		matchesType = *output.ValueReference.Type == typeName
+	} else if output.ValueReference.Reference != nil {
+		matchesType = strings.HasPrefix(*output.ValueReference.Reference, fmt.Sprintf("%s/", typeName))
```

Comment:
Could this be an external URL? Then this check doesn't work.

Maybe we should require `reference.type` for every reference, @bramwesselo ?

Reply:
I don't know how to resolve this without having the base url of the FHIR resource service present in this method. I could write a regex implementation that does something like this:

^.http://.*/(type)/.*$ 

But I can't guarantee that will work in all situations.

Reply:
Implemented something + tests, seems to work for simple `https://example.com/fhir/Questionnaire/1` references. Don't reference anything fancy.